### ISSUE TITLE
Different versions are not uploaded to S3 (Using Fog storage)

### ISSUE THREAD

**Issue by anduong** [carrierwaveuploader/carrierwave#2019]
Hi team,

I'm using Rails 4.2.3, Ruby 2.2.3 and carrierwave at Master branch.

I have 2 more versions (thumb and large) beside the original version of an image. **When I upload the image, there is only the original version on S3. The others (thumb and large) are not uploaded to S3.**

When I inspect the the image uploader, 
`Tour.find(6).images.first.versions`
it says:

``` Ruby
{:thumb=>#<ImageUploader::Uploader70188969586460:0x007fac5192d7e0 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @parent_version=#<ImageUploader:0x007fac5192eaf0 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @storage=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac5192d998 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={...}>, @storage=#<CarrierWave::Storage::Fog:0x007fac5192d6c8 @uploader=#<ImageUploader::Uploader70188969586460:0x007fac5192d7e0 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac5192cb38 @uploader=#<ImageUploader::Uploader70188969586460:0x007fac5192d7e0 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192d6c8 @uploader=#<ImageUploader::Uploader70188969586460:0x007fac5192d7e0 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/thumb_5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={}>, :view=>#<ImageUploader::Uploader70188969472620:0x007fac5192d7b8 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @parent_version=#<ImageUploader:0x007fac5192eaf0 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @storage=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac5192d998 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={...}>, @storage=#<CarrierWave::Storage::Fog:0x007fac5192c9f8 @uploader=#<ImageUploader::Uploader70188969472620:0x007fac5192d7b8 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac5192c110 @uploader=#<ImageUploader::Uploader70188969472620:0x007fac5192d7b8 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192c9f8 @uploader=#<ImageUploader::Uploader70188969472620:0x007fac5192d7b8 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/view_5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={}>, :large=>#<ImageUploader::Uploader70188969392820:0x007fac5192d790 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @parent_version=#<ImageUploader:0x007fac5192eaf0 @model=#<Tour id: 6, name: "Lapa Loca Party", description: "<p>t wouldn't be quite right to spend a Friday nig...", overview: "<p>Não seria correto passar uma noite de sexta-fei...", price: #<BigDecimal:7fac51926030,'0.18E2',9(18)>, user_id: 2, location_id: 1, duration_in_minutes: 0, pickup_address_id: 1, min_quantity: 1, max_quantity: 500, visible: true, status: "approved", slug: "lapa-loca-party", instant_booking: nil, price_type: 1, created_at: "2016-09-02 11:18:16", updated_at: "2016-09-11 06:15:58", images: ["5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg"], offerings: "", languages: ["English", "Portuguese"], flexible_price: nil, flexible_requirement: nil, thumbnail_image: "eacc73a0-067e-467e-9c20-8a4eca9a37f0.jpg", featured: false, title: nil, transportations: [], cancellation: 0, emotions: [], start_time: nil, video_presentation: nil>, @mounted_as=:images, @storage=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac5192d998 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192e9b0 @uploader=#<ImageUploader:0x007fac5192eaf0 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={...}>, @storage=#<CarrierWave::Storage::Fog:0x007fac5192ec08 @uploader=#<ImageUploader::Uploader70188969392820:0x007fac5192d790 ...>>, @file=#<CarrierWave::Storage::Fog::File:0x007fac519272f0 @uploader=#<ImageUploader::Uploader70188969392820:0x007fac5192d790 ...>, @base=#<CarrierWave::Storage::Fog:0x007fac5192ec08 @uploader=#<ImageUploader::Uploader70188969392820:0x007fac5192d790 ...>>, @path="/Users/anduong/code/tourmega/public/uploads/tour/images/6/large_5db20470-d8fd-48cc-a091-9eff7ab8afb1.jpg">, @versions={}>}
```

Does that mean the other versions are created but not uploaded to S3?

Here is my configuration for carrierwave

``` Ruby
if Rails.env.test?
    config.storage = :file
    config.enable_processing = false
  else
    config.storage = 'fog/aws'
    config.fog_credentials = {
      provider: 'AWS',
      aws_access_key_id: Figaro.env.AWS_ACCESS_KEY_ID,
      aws_secret_access_key: Figaro.env.AWS_SECRET_ACCESS_KEY,
      region: Figaro.env.AWS_REGION
    }
    config.asset_host = Figaro.env.CDN_ASSET_HOST
    config.fog_directory = Figaro.env.S3_BUCKET
    config.fog_public = true
    config.fog_authenticated_url_expiration = 900
    config.fog_attributes = { 'Cache-Control' => 'max-age=315576000' }
    config.max_file_size = 50.megabytes
  end
```

**I don't use this code to empty tmp dir** https://github.com/carrierwaveuploader/carrierwave/wiki/How-to:-Delete-cache-garbage-directories

**But I use this one to recreate versions, it works**
https://github.com/carrierwaveuploader/carrierwave/wiki/How-to%3A-Recreate-and-reprocess-your-files-stored-on-fog

Please advise. Thanks.

**Comment by mshibuya** [comment]
How do your model and uploader look like?

**Comment by anduong** [comment]
@mshibuya 

Here is my uploader

```Ruby
# encoding: utf-8

class ImageUploader < BaseUploader

  # Include RMagick or MiniMagick support:
  # include CarrierWave::RMagick
  include CarrierWave::MiniMagick

  # Choose what kind of storage to use for this uploader:
  # storage :file

  # Override the directory where uploaded files will be stored.
  # This is a sensible default for uploaders that are meant to be mounted:
  # Process files as they are uploaded:
  # process :scale => [200, 300]
  #
  # def scale(width, height)
  #   # do something
  #   # process :convert => 'png'
  # end

  process :optimize
  # process :store_dimensions
  process :convert => 'jpg'

  # Create different versions of your uploaded files:
  version :thumb do
    process resize_to_fill: [355, 235]
  end

  version :view do
    process resize_to_fill: [600, 400]
  end

  version :large do
    process resize_to_fill: [800, 600]
    process :optimize
  end
  # version :medium, :from_version => :large do
  #   process resize_to_fill: [299, 176]
  # end

  # version :medium_home, :from_version => :large do
  #   process resize_to_fill: [370, 218]
  # end

  # version :large, :from_version => :xlarge do
  #   process resize_to_fill: [460, 292]
  # end

  # version :xlarge do
  #   process resize_to_fill: [620, 364]
  # end

  # version :view do
  #   process resize_to_fill_if_larger: [645, 318]
  # end

  # version :view_larger do
  #   process resize_to_fill_if_larger: [645, 430]
  # end
  def image
    @image ||= MiniMagick::Image.open( model.send(mounted_as).path )
  end

  def image_width
     image[:width]
  end

  def image_height
    image[:height]
  end

  def resize_to_fill_if_larger(width, height, gravity='center')
    width = self.image_width if self.image_width < width
    height = self.image_height if self.image_height < height
    resize_to_fill(width, height, gravity)
  end

  def optimize
    manipulate! do |img|
      return img unless img.mime_type.match /image\/jpeg/
      img.strip
      img.combine_options do |c|
        c.quality "80"
        c.depth "8"
        c.interlace "plane"
      end
      img
    end
  end

  # for image size validation
  # fetching dimensions in uploader, validating it in model
  attr_accessor :width, :height
  before :cache, :capture_size
  def capture_size(file)
    if version_name.blank? # Only do this once, to the original version
      if file.path.nil? # file sometimes is in memory
        img = ::MiniMagick::Image::read(file.file)
        @width = img[:width]
        @height = img[:height]
      else
        @width, @height = `identify -format "%wx %h" #{file.path}`.split(/x/).map{|dim| dim.to_i }
      end
    end
  end

  # Add a white list of extensions which are allowed to be uploaded.
  # For images you might use something like this:
  def extension_white_list
    %w(jpg jpeg gif png)
  end

  def filename
    if original_filename.present?
      @updated_name ||= "#{SecureRandom.uuid}.jpg"
    end
  end

  def store_dimensions
    if file && model
      model.width, model.height = ::MiniMagick::Image.open(file.file)[:dimensions]
    end
  end

  def fog_public
    true
  end

  def asset_host
    ENV['CDN_ASSET_HOST']
  end
end

```

Here is base uploader

```Ruby
# encoding: utf-8

class BaseUploader < CarrierWave::Uploader::Base
  include CarrierWave::MiniMagick

  # Choose what kind of storage to use for this uploader:
  if Rails.env.test?
    storage :file
  else
    storage :fog
  end
  # Override the directory where uploaded files will be stored.
  # This is a sensible default for uploaders that are meant to be mounted:
  def store_dir
    "#{Rails.root}/public/uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end

  # def cache_dir
  #   "#{Rails.root}/tmp/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  # end

  before :store, :remember_cache_id
  after :store, :delete_tmp_dir
   # store! nil's the cache_id after it finishes so we need to remember it for deletion
  def remember_cache_id(new_file)
    @cache_id_was = cache_id
  end

  def delete_tmp_dir(new_file)
    # make sure we don't delete other things accidentally by checking the name pattern
    if @cache_id_was.present? && @cache_id_was =~ /\A[\d]{8}\-[\d]{4}\-[\d]+\-[\d]{4}\z/
      FileUtils.rm_rf(File.join(root, cache_dir, @cache_id_was))
    end
  end
end

```

But ignore the following code, because I commented it out. But still did not work :(
```Ruby
 before :store, :remember_cache_id
  after :store, :delete_tmp_dir
   # store! nil's the cache_id after it finishes so we need to remember it for deletion
  def remember_cache_id(new_file)
    @cache_id_was = cache_id
  end

  def delete_tmp_dir(new_file)
    # make sure we don't delete other things accidentally by checking the name pattern
    if @cache_id_was.present? && @cache_id_was =~ /\A[\d]{8}\-[\d]{4}\-[\d]+\-[\d]{4}\z/
      FileUtils.rm_rf(File.join(root, cache_dir, @cache_id_was))
    end
  end
end
```

**Comment by mshibuya** [comment]
Why are you so sure that `remember_cache_id` and `delete_tmp_dir` are not related to the issue?
'Special Consideration' section clearly states that it does not work when the uploader has multiple version.
https://github.com/carrierwaveuploader/carrierwave/wiki/How-to:-Delete-cache-garbage-directories

**Comment by anduong** [comment]
@mshibuya Thanks for your response. I have tried commenting that part of code out but it did not work. But let me try commenting it out one more time : )

**Comment by anduong** [comment]
@mshibuya I tried to comment the code out again and checked the images on S3. They were in different names because of this snippet code
```Ruby
  def filename
    if original_filename.present?
      @updated_name ||= "#{SecureRandom.uuid}.jpg"
    end
  end
```

This is the reason why last time I only saw the original image with name for example: **abc.png** but did not see thumb version of it which should be **thumb_abc.png**

The above snippet code ruined all the name convention of different versions. I think this one would help https://github.com/carrierwaveuploader/carrierwave/wiki/How-to:-Customize-your-version-file-names

Thanks for your support. This issue can be closed.
